File: Graph(그래프)/DFS(깊이 우선 탐색)/백준_4963_섬의갯수/섬의 갯수.cs
using System.Collections;
using System.Collections.Generic;
using static System.Console;
using System;

class Program
{
    static bool[][] arr = new bool[52][];		//초기화가 true, false로	
	static void DFS(int i, int j)
	{
		if (arr[i][j] == false) return;

		arr[i][j] = false;
		DFS(i + 1, j);
		DFS(i + 1, j - 1);
		DFS(i + 1, j + 1);
		DFS(i, j - 1);
		DFS(i, j + 1);
		DFS(i - 1, j);
		DFS(i - 1, j - 1);
		DFS(i - 1, j + 1);
	}
	public static void Main(string[] args)
    {
		int w, h;
		for(int i = 0; i< arr.Length; i++)
			arr[i] = new bool[52];

		while (true)
        {
			string[] input = ReadLine().Split(' ');
			w = int.Parse(input[0]);
			h = int.Parse(input[1]);
			if (w == 0 && h == 0) break;
			int count = 0;
			for(int i = 1; i<=h; i++)
            {
				input = ReadLine().Split(' ');
				for(int j = 1; j<=w; j++)
                {
					if (int.Parse(input[j - 1]) == 1) arr[i][j] = true;
					else arr[i][j] = false;
                }
            }

			for(int i = 1; i<=h; i++)
            {
				for(int j = 1; j<= w; j++)
                {
                    if (arr[i][j]) {
						count++;
						DFS(i, j);
					}
                }
            }
			WriteLine(count);
        }
    }
}

File: Graph(그래프)/DFS(깊이 우선 탐색)/백준_4963_섬의갯수/섬의 갯수.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

bool arr[51][51];

void DFS(int i, int j) {
	if (arr[i][j] == false) return;

	arr[i][j] = false;
	DFS(i + 1, j);
	DFS(i + 1, j - 1);
	DFS(i + 1, j + 1);
	DFS(i, j - 1);
	DFS(i, j + 1);
	DFS(i - 1, j);
	DFS(i - 1, j - 1);
	DFS(i - 1, j + 1);
}

int main() {
	int w;
	int h;
	
	while (1) {
		cin >> w >> h;
		if (w == 0 && h == 0) break;
		int count = 0;
		for (int i = 1; i <= h; i++) {
			for (int j = 1; j <= w; j++) {
				cin >> arr[i][j];
			}
		}

		for (int i = 1; i <= h; i++) {
			for (int j = 1; j <= w; j++) {
				if (arr[i][j]) {
					count++;
					DFS(i, j);
				}
			}
		}
		
		cout << count << '\n';
	}


}